### ISSUE TITLE
[10회차]함수화 및 서버배포

### ISSUE THREAD

**Issue by smartkai** [smartkai/spot#13]
### 클라우드 서버
1. 작업 URL : https://goor.me/4qTgcecCh1Tg9tDw8
   --> root 권한부여
2. 사용자 URL : https://spotcop.run.goorm.site/

### 학습모델
1. LSTM 예측모델 참고용 코랩 : https://colab.research.google.com/drive/1PxXlnjwSa1Awb2g4Aaqws9LZJDBnIIXi
2. GRU 방식추천 코랩(완성) : https://colab.research.google.com/drive/1bBvPIr3gY_lAnZSvNc1MyYm4JUbNPuAl
   --> 프로젝트 채택!!

### RED TEAM (강주연)
1. [백앤드] 파이썬파일 : /workspace/spot/app/home/red.py
  --> 기존 샘플형식 input/output 수정금지
4. [프론트] 화면파일 : /workspace/spot/app/templates/redDetail.html

### BLUE TEAM (허유정)
1. [백앤드] 파이썬파일 : /workspace/spot/app/home/blue.py
  --> 기존 샘플형식 input/output 수정금지
2. [프론트] 화면파일 : /workspace/spot/app/templates/blueDetail.html

### 참고 (입/출력 형식 표준화)
- 데이터 입력값은 : 지역,
- 데이터 출력값은 : 일자+거래량+금액
- 예시형식
- ========== json_data_month ======== 
[{'deal_date': '2023-11-01', 'mean': 7965, 'count': 115}, {'deal_date': '2023-12-01', 'mean': 8104, 'count': 100}, {
'deal_date': '2024-01-01', 'mean': 8453, 'count': 193}, {'deal_date': '2024-02-01', 'mean': 8786, 'count': 162}, {'deal_date': '2024-03-01', 'mean': 8470
, 'count': 220}, {'deal_date': '2024-04-01', 'mean': 9091, 'count': 225}, {'deal_date': '2024-05-01', 'mean': 8800, 'count': 115}, {'deal_date': '2024-06
-01', 'mean': 8896, 'count': 280}, {'deal_date': '2024-07-01', 'mean': 8078, 'count': 10}, {'deal_date': '2024-08-01', 'mean': 10953, 'count': 10}, {'dea
l_date': '2024-09-01', 'mean': 10967, 'count': 10}, {'deal_date': '2024-10-01', 'mean': 10327, 'count': 10}]

**Comment by smartkai** [comment]
import pandas as pd
import numpy as np
from sklearn.preprocessing import MinMaxScaler
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import LSTM, Dense, Dropout
from tensorflow.keras.callbacks import EarlyStopping

def predict_real_estate(data, features, target, future_days):
    """
    주어진 데이터로 미래 거래량과 금액을 예측하는 함수
    
    Parameters:
        data (DataFrame): 입력 데이터 (필수 열: '매매거래일', features, target)
        features (list): 학습에 사용할 특성 리스트
        target (list): 예측 대상 열 리스트
        future_days (int): 예측하고자 하는 미래 일수
        
    Returns:
        list: 미래 거래량과 매매 금액 예측값 리스트
    """
    # 날짜 정렬
    data['매매거래일'] = pd.to_datetime(data['매매거래일'])
    data = data.sort_values(by='매매거래일')
    
    # 데이터 스케일링
    scaler = MinMaxScaler()
    scaled_data = scaler.fit_transform(data[features + target])
    
    # 윈도우 생성
    def create_sequences(data, seq_length):
        x = []
        y = []
        for i in range(len(data) - seq_length):
            x.append(data[i:i + seq_length, :-len(target)])
            y.append(data[i + seq_length, -len(target):])
        return np.array(x), np.array(y)
    
    sequence_length = 30  # 과거 30일 데이터를 사용
    x, y = create_sequences(scaled_data, sequence_length)
    
    # 데이터 분리
    train_size = int(len(x) * 0.8)
    x_train, x_test = x[:train_size], x[train_size:]
    y_train, y_test = y[:train_size], y[train_size:]
    
    # 모델 생성
    model = Sequential([
        LSTM(64, return_sequences=True, input_shape=(x_train.shape[1], x_train.shape[2])),
        Dropout(0.2),
        LSTM(32),
        Dropout(0.2),
        Dense(len(target))
    ])
    model.compile(optimizer='adam', loss='mse')
    
    # 모델 학습
    early_stop = EarlyStopping(monitor='val_loss', patience=10, restore_best_weights=True)
    model.fit(
        x_train, y_train,
        validation_data=(x_test, y_test),
        epochs=100,
        batch_size=32,
        callbacks=[early_stop],
        verbose=0  # 학습 로그 출력 억제
    )
    
    # 미래 예측
    predictions = []
    future_input = scaled_data[-sequence_length:, :-len(target)]  # 최근 30일 데이터
    
    for _ in range(future_days):
        future_input = np.expand_dims(future_input, axis=0)
        future_prediction = model.predict(future_input)[0]  # 모델 예측
        predictions.append(future_prediction)
        
        # 예측값을 기존 입력 데이터에 추가해 업데이트
        future_input = np.append(future_input[0, 1:, :], [future_prediction], axis=0)
    
    # 결과 디스케일링
    predictions = np.array(predictions)
    predictions = scaler.inverse_transform(
        np.hstack((np.zeros((predictions.shape[0], len(features))), predictions))
    )[:, -len(target):]
    
    return predictions

# 샘플 데이터
data = pd.read_csv('real_estate_data.csv')  # 파일명
features = ['금리', '환율', '코스피', '매매거래량']  # 입력 특성
target = ['매매가격', '매매거래량']  # 예측 대상

# 미래 7일 예측
future_days = 7
predictions = predict_real_estate(data, features, target, future_days)

# 결과 출력
for i, prediction in enumerate(predictions, 1):
    print(f"미래 {i}일 후: 매매가격: {prediction[0]:,.2f}, 거래량: {prediction[1]:,.2f}")

**Comment by smartkai** [comment]
아래는 함수를 사용하여 미래 일자 입력 시 해당 날짜의 거래량과 금액을 예측하는 코드입니다. 이 함수는 모델 생성, 학습, 예측 단계를 포함합니다.

코드 구현

import pandas as pd
import numpy as np
from sklearn.preprocessing import MinMaxScaler
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import LSTM, Dense, Dropout
from tensorflow.keras.callbacks import EarlyStopping

def predict_real_estate(data, features, target, future_days):
    """
    주어진 데이터로 미래 거래량과 금액을 예측하는 함수
    
    Parameters:
        data (DataFrame): 입력 데이터 (필수 열: '매매거래일', features, target)
        features (list): 학습에 사용할 특성 리스트
        target (list): 예측 대상 열 리스트
        future_days (int): 예측하고자 하는 미래 일수
        
    Returns:
        list: 미래 거래량과 매매 금액 예측값 리스트
    """
    # 날짜 정렬
    data['매매거래일'] = pd.to_datetime(data['매매거래일'])
    data = data.sort_values(by='매매거래일')
    
    # 데이터 스케일링
    scaler = MinMaxScaler()
    scaled_data = scaler.fit_transform(data[features + target])
    
    # 윈도우 생성
    def create_sequences(data, seq_length):
        x = []
        y = []
        for i in range(len(data) - seq_length):
            x.append(data[i:i + seq_length, :-len(target)])
            y.append(data[i + seq_length, -len(target):])
        return np.array(x), np.array(y)
    
    sequence_length = 30  # 과거 30일 데이터를 사용
    x, y = create_sequences(scaled_data, sequence_length)
    
    # 데이터 분리
    train_size = int(len(x) * 0.8)
    x_train, x_test = x[:train_size], x[train_size:]
    y_train, y_test = y[:train_size], y[train_size:]
    
    # 모델 생성
    model = Sequential([
        LSTM(64, return_sequences=True, input_shape=(x_train.shape[1], x_train.shape[2])),
        Dropout(0.2),
        LSTM(32),
        Dropout(0.2),
        Dense(len(target))
    ])
    model.compile(optimizer='adam', loss='mse')
    
    # 모델 학습
    early_stop = EarlyStopping(monitor='val_loss', patience=10, restore_best_weights=True)
    model.fit(
        x_train, y_train,
        validation_data=(x_test, y_test),
        epochs=100,
        batch_size=32,
        callbacks=[early_stop],
        verbose=0  # 학습 로그 출력 억제
    )
    
    # 미래 예측
    predictions = []
    future_input = scaled_data[-sequence_length:, :-len(target)]  # 최근 30일 데이터
    
    for _ in range(future_days):
        future_input = np.expand_dims(future_input, axis=0)
        future_prediction = model.predict(future_input)[0]  # 모델 예측
        predictions.append(future_prediction)
        
        # 예측값을 기존 입력 데이터에 추가해 업데이트
        future_input = np.append(future_input[0, 1:, :], [future_prediction], axis=0)
    
    # 결과 디스케일링
    predictions = np.array(predictions)
    predictions = scaler.inverse_transform(
        np.hstack((np.zeros((predictions.shape[0], len(features))), predictions))
    )[:, -len(target):]
    
    return predictions

# 샘플 데이터
data = pd.read_csv('real_estate_data.csv')  # 파일명
features = ['금리', '환율', '코스피', '매매거래량']  # 입력 특성
target = ['매매가격', '매매거래량']  # 예측 대상

# 미래 7일 예측
future_days = 7
predictions = predict_real_estate(data, features, target, future_days)

# 결과 출력
for i, prediction in enumerate(predictions, 1):
    print(f"미래 {i}일 후: 매매가격: {prediction[0]:,.2f}, 거래량: {prediction[1]:,.2f}")

주요 특징

	1.	함수화: predict_real_estate 함수는 입력 데이터와 특성, 목표 열, 미래 예측 기간을 받아 예측 결과를 반환합니다.
	2.	유연성: 입력 데이터와 미래 예측 일수를 쉽게 조정 가능.
	3.	결과 디스케일링: 모델이 예측한 값을 원래 단위로 변환해 출력.
	4.	미래 입력 업데이트: 각 예측 결과를 다음 입력으로 사용해 시계열 데이터를 확장.

예상 출력

예를 들어, future_days=7로 설정하면 미래 7일간의 매매가격과 거래량이 출력됩니다.

미래 1일 후: 매매가격: 123,456.78, 거래량: 567.89
미래 2일 후: 매매가격: 124,678.12, 거래량: 578.34
...

이 코드를 실행하려면, 적절한 데이터 파일과 특성을 지정해야 합니다. 데이터 준비와 관련한 질문이 있다면 알려주세요!

**Comment by smartkai** [comment]
import pandas as pd
import numpy as np
from sklearn.preprocessing import MinMaxScaler
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import LSTM, Dense, Dropout
from tensorflow.keras.callbacks import EarlyStopping

def predict_real_estate(data, features, target, future_days):
    """
    주어진 데이터로 미래 거래량과 금액을 예측하는 함수
    
    Parameters:
        data (DataFrame): 입력 데이터 (필수 열: '매매거래일', features, target)
        features (list): 학습에 사용할 특성 리스트
        target (list): 예측 대상 열 리스트
        future_days (int): 예측하고자 하는 미래 일수
        
    Returns:
        list: 미래 거래량과 매매 금액 예측값 리스트
    """
    # 날짜 정렬
    data['매매거래일'] = pd.to_datetime(data['매매거래일'])
    data = data.sort_values(by='매매거래일')
    
    # 데이터 스케일링
    scaler = MinMaxScaler()
    scaled_data = scaler.fit_transform(data[features + target])
    
    # 윈도우 생성
    def create_sequences(data, seq_length):
        x = []
        y = []
        for i in range(len(data) - seq_length):
            x.append(data[i:i + seq_length, :-len(target)])
            y.append(data[i + seq_length, -len(target):])
        return np.array(x), np.array(y)
    
    sequence_length = 30  # 과거 30일 데이터를 사용
    x, y = create_sequences(scaled_data, sequence_length)
    
    # 데이터 분리
    train_size = int(len(x) * 0.8)
    x_train, x_test = x[:train_size], x[train_size:]
    y_train, y_test = y[:train_size], y[train_size:]
    
    # 모델 생성
    model = Sequential([
        LSTM(64, return_sequences=True, input_shape=(x_train.shape[1], x_train.shape[2])),
        Dropout(0.2),
        LSTM(32),
        Dropout(0.2),
        Dense(len(target))
    ])
    model.compile(optimizer='adam', loss='mse')
    
    # 모델 학습
    early_stop = EarlyStopping(monitor='val_loss', patience=10, restore_best_weights=True)
    model.fit(
        x_train, y_train,
        validation_data=(x_test, y_test),
        epochs=100,
        batch_size=32,
        callbacks=[early_stop],
        verbose=0  # 학습 로그 출력 억제
    )
    
    # 미래 예측
    predictions = []
    future_input = scaled_data[-sequence_length:, :-len(target)]  # 최근 30일 데이터
    
    for _ in range(future_days):
        future_input = np.expand_dims(future_input, axis=0)
        future_prediction = model.predict(future_input)[0]  # 모델 예측
        predictions.append(future_prediction)
        
        # 예측값을 기존 입력 데이터에 추가해 업데이트
        future_input = np.append(future_input[0, 1:, :], [future_prediction], axis=0)
    
    # 결과 디스케일링
    predictions = np.array(predictions)
    predictions = scaler.inverse_transform(
        np.hstack((np.zeros((predictions.shape[0], len(features))), predictions))
    )[:, -len(target):]
    
    return predictions

# 샘플 데이터
data = pd.read_csv('real_estate_data.csv')  # 파일명
features = ['금리', '환율', '코스피', '매매거래량']  # 입력 특성
target = ['매매가격', '매매거래량']  # 예측 대상

# 미래 7일 예측
future_days = 7
predictions = predict_real_estate(data, features, target, future_days)

# 결과 출력
for i, prediction in enumerate(predictions, 1):
    print(f"미래 {i}일 후: 매매가격: {prediction[0]:,.2f}, 거래량: {prediction[1]:,.2f}")

**Comment by kangjucop** [comment]
https://colab.research.google.com/drive/1W3OkpTM00eI6wgrabvWpuHOuw51W38vq#scrollTo=umUMk7wJ_gAE

**Comment by hohohouuuuuuui** [comment]
(1) 블루팀 코랩 - 선재만, 백숙진, 허유정
https://colab.research.google.com/drive/1fyRU8rjeHSc-iSWMcqjIqXEKQF-G-Xc5?usp=sharing

(2) groom 파일 변경
-blue.py, views.py 
-거시지수 항목 추가, 모델학습파일 저장

**Comment by smartkai** [comment]
### 최종 features  내역

features = ["아파트매매", "아파트전세", "기준금리", "시장금리(10년국고채)", "코스피 거래대금", "소비자물가지수", "가계대출", "본원통화", "count", "median"]

**Comment by smartkai** [comment]
def ecos_api_green(tail):
    # tail 예시 = "901Y009/M/201501/202412/0"
    url = f'http://ecos.bok.or.kr/api/StatisticSearch/{api_key}/json/kr/1/1000/' + tail
    r = requests.get(url)
    jo = r.json()
    df2 = pd.DataFrame(jo['StatisticSearch']['row'])
    df2 = df2[['TIME', 'DATA_VALUE']]
    df2['deal_date'] = pd.to_datetime(df2['TIME'], format='%Y%m')
    df2['DATA_VALUE'] = pd.to_numeric(df2['DATA_VALUE'])
    df2 = df2.drop(columns=['TIME'])
    df2 = df2.set_index('deal_date')
    return df2
	
apt_pay = ecos_api_green('901Y062/M/202311/202410/P63ACA') # 주택매매가격지수(KB) > 아파트(서울)

**Comment by smartkai** [comment]
df2['아파트매매'] = apt_pay['DATA_VALUE']

features = ["아파트매매", "아파트전세", "기준금리", "시장금리(10년국고채)", "코스피 거래대금", "소비자물가지수", "가계대출", "본원통화", "count", "median"]

**Comment by smartkai** [comment]
# df2에 추가할 데이터 불러오기
df2['아파트매매'] = apt_pay['DATA_VALUE']

# 1. 주택담보대출 금리
housing_loan_rate = ecos_api_green('721Y001/M/202311/202410/6010101')
df2['주택담보대출금리'] = housing_loan_rate['DATA_VALUE']

# 2. 가계저축률
household_savings_rate = ecos_api_green('901Y006/M/202311/202410/5000001')
df2['가계저축률'] = household_savings_rate['DATA_VALUE']

# 3. 주택구입부담지수
housing_affordability = ecos_api_green('901Y027/Q/201001/202410/4300201')
df2['주택구입부담지수'] = housing_affordability['DATA_VALUE']

# 4. 국내총생산(GDP) 성장률
gdp_growth_rate = ecos_api_green('111Y002/Q/201001/202410/1010101')
df2['GDP성장률'] = gdp_growth_rate['DATA_VALUE']

# 5. 소비자신뢰지수(CSI)
consumer_confidence = ecos_api_green('731Y004/M/202311/202410/1000000')
df2['소비자신뢰지수'] = consumer_confidence['DATA_VALUE']

# 6. 실업률
unemployment_rate = ecos_api_green('901Y010/M/201001/202410/8000000')
df2['실업률'] = unemployment_rate['DATA_VALUE']

# 7. 전세가율
jeonse_ratio = ecos_api_green('901Y062/M/202311/202410/P63ACA')
df2['전세가율'] = jeonse_ratio['DATA_VALUE']

# 8. 지역별 주택공급량
housing_supply = ecos_api_green('901Y063/M/201501/202412/P63ACD')
df2['주택공급량'] = housing_supply['DATA_VALUE']

# 9. 건축허가 건수
construction_permits = ecos_api_green('901Y035/M/201001/202410/5300000')
df2['건축허가건수'] = construction_permits['DATA_VALUE']

# 10. 환율 (KRW/USD)
exchange_rate = ecos_api_green('731Y002/M/201001/202410/1030000')
df2['환율'] = exchange_rate['DATA_VALUE']

# 11. 주택담보대출 잔액 증가율
housing_loan_balance = ecos_api_green('901Y001/M/201001/202410/4010201')
df2['주택담보대출잔액증가율'] = housing_loan_balance['DATA_VALUE']

# 12. 미분양 주택 수
unsold_houses = ecos_api_green('901Y066/M/201001/202410/P63ACC')
df2['미분양주택수'] = unsold_houses['DATA_VALUE']

# 13. 지역별 인구 이동
population_movement = ecos_api_green('901Y025/M/201001/202410/2000000')
df2['인구이동'] = population_movement['DATA_VALUE']

# Feature 리스트에 추가
features = [
    "아파트매매", 
    "아파트전세", 
    "기준금리", 
    "시장금리(10년국고채)", 
    "코스피 거래대금", 
    "소비자물가지수", 
    "가계대출", 
    "본원통화", 
    "주택담보대출금리", 
    "가계저축률", 
    "주택구입부담지수", 
    "GDP성장률", 
    "소비자신뢰지수", 
    "실업률", 
    "전세가율", 
    "주택공급량", 
    "건축허가건수", 
    "환율", 
    "주택담보대출잔액증가율", 
    "미분양주택수", 
    "인구이동", 
    "count", 
    "median"
]
